# Stephen Kay <br> #
# CSCI-3287 Course Project <br> #
# Gaming the NBA<br> #

In [1]:
%load_ext sql
%matplotlib inline

In [2]:
import matplotlib.pyplot as plt
import zipfile
import pandas as pd
from sqlalchemy import create_engine
import sqlalchemy.sql

In [3]:
import getpass
# passwd = getpass.getpass()
passwd = 'M6f9NYDl6FgaBMd1jyLBxpAYVXHcBXl3'
dbuser='qktewplm'
eng = create_engine('postgresql://{0}:{1}@otto.db.elephantsql.com:5432/qktewplm'.format(dbuser, passwd))
con = eng.connect()

In [4]:
con.execute("SELECT postgis_full_version()").fetchall()

[('POSTGIS="2.5.1 r17027" [EXTENSION] (liblwgeom version mismatch: "2.5.3 r17699") PGSQL="110" GEOS="3.6.2-CAPI-1.10.2 4d2925d6" PROJ="Rel. 4.9.3, 15 August 2016" GDAL="GDAL 2.2.3, released 2017/11/20" LIBXML="2.9.4" LIBJSON="0.12.1" LIBPROTOBUF="1.2.1" RASTER',)]

In [5]:
rs = con.execute('''
DROP TABLE IF EXISTS nba_schedule;
CREATE TABLE nba_schedule
	(
	  id numeric,
	  road_team character varying(50),
	  home_team character varying(50),
	  total_game_minutes numeric,
	  road_team_final_score numeric,
	  home_team_final_score numeric,
      road_team_rest character varying(50),
      home_team_rest character varying(50)
	);
''')

In [6]:
con.execute("SELECT * from nba_schedule LIMIT 1").fetchall()

[]

## Ok now let's use the wonderful Pandas to read in our CSV and convert it into a dictionary! ##

In [7]:
nba_data = pd.read_csv("2017-2018_NBA_Historical_Schedule.csv")

In [8]:
nba_data.head()

id  road_team     home_team  total_game_minutes  \
0  21700001     Boston     Cleveland               240.0   
1  21700002    Houston  Golden State               240.0   
2  21700003  Charlotte       Detroit               240.0   
3  21700004   Brooklyn       Indiana               240.0   
4  21700005      Miami       Orlando               240.0   

   road_team_final_score  home_team_final_score road_team_rest home_team_rest  
0                     99                    102             3+             3+  
1                    122                    121             3+             3+  
2                     90                    102             3+             3+  
3                    131                    140             3+             3+  
4                    109                    116             3+             3+

In [9]:
nba_dict = nba_data.to_dict('records')

In [10]:
nba_dict[0]

{'id': 21700001,
 'road_team': 'Boston',
 'home_team': 'Cleveland',
 'total_game_minutes': 240.0,
 'road_team_final_score': 99,
 'home_team_final_score': 102,
 'road_team_rest': '3+',
 'home_team_rest': '3+'}

In [11]:
for row in nba_dict:
    cmd = sqlalchemy.sql.text('''INSERT INTO nba_schedule(id, road_team, home_team, total_game_minutes,
    road_team_final_score, home_team_final_score, road_team_rest, home_team_rest)\
    VALUES ( :id, :road_team, :home_team, :total_game_minutes, :road_team_final_score,
    :home_team_final_score, :road_team_rest, :home_team_rest) ''' )
    con.execute(cmd, row)

In [12]:
con.execute("SELECT * from nba_schedule order by id asc LIMIT 1").fetchall()

[(Decimal('21700001'), 'Boston', 'Cleveland', Decimal('240.0'), Decimal('99'), Decimal('102'), '3+', '3+')]

In [13]:
nba_dict[0]

{'id': 21700001,
 'road_team': 'Boston',
 'home_team': 'Cleveland',
 'total_game_minutes': 240.0,
 'road_team_final_score': 99,
 'home_team_final_score': 102,
 'road_team_rest': '3+',
 'home_team_rest': '3+'}

In [14]:
data_list_len = len(nba_data)
nba_dict[data_list_len-1]

{'id': 41700404,
 'road_team': 'Golden State',
 'home_team': 'Cleveland',
 'total_game_minutes': 240.0,
 'road_team_final_score': 108,
 'home_team_final_score': 85,
 'road_team_rest': '1',
 'home_team_rest': '1'}

The above SQL and dictionary statements ensure that my database has all of the rows by checking the database first and last entries against the dictionaries' view of the first and last entries.

In [15]:
import geopandas

In [16]:
nba_cities_data = pd.read_csv("NBA-Cities-Lat-Long.csv")

In [17]:
nba_cities_data_dict = nba_cities_data.to_dict('records')

In [18]:
from shapely.geometry import Point

nba_cities_data['geometry'] = list(zip(nba_cities_data.lon, nba_cities_data.lat))
nba_cities_data['geometry'] = nba_cities_data['geometry'].apply(Point)

In [19]:
rs = con.execute('''
DROP TABLE IF EXISTS nba_cities;
CREATE TABLE nba_cities
	(
	  city character varying(50),
      lat numeric,
      lon numeric
	);
''')

In [20]:
for row in nba_cities_data_dict:
    cmd = sqlalchemy.sql.text('''INSERT INTO nba_cities(city, lat, lon)\
    VALUES (:city, :lat, :lon) ''' )
    con.execute(cmd, row)

In [21]:
con.execute("SELECT * from nba_cities LIMIT 1").fetchall()

[('Atlanta', Decimal('33.749'), Decimal('94.38799999999999'))]

Ok this looks good! Now let's make sure that we have a GEOGRAPHY column to work with by zipping together the lats and longs. This will come in handy later when deciding if travel distance affects how a team play on a given night!

In [333]:
con.execute('''ALTER TABLE nba_cities DROP COLUMN IF EXISTS geog;
               ALTER TABLE nba_cities ADD COLUMN geog GEOGRAPHY(POINT,4326)''')

In [334]:
con.execute("UPDATE nba_cities SET geog = ST_SetSRID(ST_MakePoint(lon,lat),4326);")

In [335]:
con.execute("SELECT * from nba_cities LIMIT 1").fetchall()

[('Atlanta', Decimal('33.749'), Decimal('94.38799999999999'), '0101000020E610000045B6F3FDD49857401D5A643BDFDF4040')]

Looks like our nba_cities table now has a GEOGRAPHY column. Excellent!

This query below is just a sanity check for how many games ouccred in the 2017-2018 NBA Season and to make sure that my table loaded all of the correct data.

In [336]:
game_total = con.execute("SELECT COUNT(*) from nba_schedule").fetchall()
print("There were, ", game_total, "total games in the 2017-2018 NBA season")

There were,  [(1312,)] total games in the 2017-2018 NBA season


## Ok now that we have 2 initial table set up let's run some simple queries to get a handle on what we can begin to glean from the data ##

### <center>Some possible queries of interest are</center> ###

<center><li> What was the average score across all games for the road and home teams in the 2017-2018 season?</li><center>

<center><li> How many games in the 2017-2018 NBA season did road teams win when they were on a back-to-back or a 3IN4 and the home team had at least 1 or more nights of rest?</li><center>

<center><li> How many games in the 2017-2018 NBA season did home teams win when they had 1 or more nights of rest and the road team was on a B2B or a 3IN4, what?</li><center>

In [33]:
con.execute('''
SELECT AVG(home_team_final_score)
    FROM nba_schedule
LIMIT 5;
''').fetchall()

[(Decimal('107.3986280487804878'),)]

In [34]:
con.execute('''
SELECT AVG(road_team_final_score)
    FROM nba_schedule
LIMIT 5;
''').fetchall()

[(Decimal('105.0320121951219512'),)]

In [35]:
### How many games in the 2017-2018 NBA season did road teams win when they were on a
### back-to-back and the home team had 1 night of rest
con.execute('''
SELECT home_team, road_team, home_team_final_score, road_team_final_score, road_team_rest, home_team_rest
    FROM nba_schedule
    WHERE road_team_rest = 'B2B'
    AND home_team_rest = '1'
    AND (road_team_final_score > home_team_final_score)
LIMIT 5;
''').fetchall()

[('Phoenix', 'LA Lakers', Decimal('130'), Decimal('132'), 'B2B', '1'),
 ('Golden State', 'Detroit', Decimal('107'), Decimal('115'), 'B2B', '1'),
 ('LA Lakers', 'Phoenix', Decimal('113'), Decimal('122'), 'B2B', '1'),
 ('LA Lakers', 'Memphis', Decimal('99'), Decimal('109'), 'B2B', '1'),
 ('Orlando', 'Miami', Decimal('111'), Decimal('117'), 'B2B', '1')]

In [36]:
###Query Total
print(len(con.execute('''
SELECT home_team, road_team, home_team_final_score, road_team_final_score, road_team_rest, home_team_rest
    FROM nba_schedule
    WHERE road_team_rest = 'B2B'
    AND home_team_rest = '1'
    AND (road_team_final_score > home_team_final_score)
''').fetchall()))

15


In [37]:
### How many games in the 2017-2018 NBA season did home teams win when they had 1 night
### of rest and the road team was on a B2B
con.execute('''
SELECT home_team, road_team, home_team_final_score, road_team_final_score, road_team_rest, home_team_rest
    FROM nba_schedule
    WHERE road_team_rest = 'B2B'
    AND home_team_rest = '1'
    AND (road_team_final_score < home_team_final_score)
LIMIT 5;
''').fetchall()

[('Phoenix', 'Utah', Decimal('97'), Decimal('88'), 'B2B', '1'),
 ('Washington', 'LA Lakers', Decimal('111'), Decimal('95'), 'B2B', '1'),
 ('Washington', 'Atlanta', Decimal('113'), Decimal('94'), 'B2B', '1'),
 ('New Orleans', 'LA Clippers', Decimal('111'), Decimal('103'), 'B2B', '1'),
 ('Minnesota', 'San Antonio', Decimal('98'), Decimal('86'), 'B2B', '1')]

In [38]:
###Query Total
print(len(con.execute('''
SELECT home_team, road_team, home_team_final_score, road_team_final_score, road_team_rest, home_team_rest
    FROM nba_schedule
    WHERE road_team_rest = 'B2B'
    AND home_team_rest = '1'
    AND (home_team_final_score > road_team_final_score)
''').fetchall()))

31


In [39]:
### How many games in the 2017-2018 NBA season did road teams win when they were on a
### back-to-back and the home team had 2 nights of rest
con.execute('''
SELECT home_team, road_team, home_team_final_score, road_team_final_score, road_team_rest, home_team_rest
    FROM nba_schedule
    WHERE road_team_rest = 'B2B'
    AND home_team_rest = '2'
    AND (road_team_final_score > home_team_final_score)
LIMIT 5;
''').fetchall()

[('Washington', 'Phoenix', Decimal('116'), Decimal('122'), 'B2B', '2'),
 ('Portland', 'Washington', Decimal('92'), Decimal('106'), 'B2B', '2'),
 ('Brooklyn', 'New Orleans', Decimal('128'), Decimal('138'), 'B2B', '2'),
 ('Phoenix', 'Denver', Decimal('113'), Decimal('123'), 'B2B', '2'),
 ('Boston', 'Washington', Decimal('124'), Decimal('125'), 'B2B', '2')]

In [40]:
###Query Total
print(len(con.execute('''
SELECT home_team, road_team, home_team_final_score, road_team_final_score, road_team_rest, home_team_rest
    FROM nba_schedule
    WHERE road_team_rest = 'B2B'
    AND home_team_rest = '2'
    AND (road_team_final_score > home_team_final_score)
''').fetchall()))

5


In [41]:
### How many games in the 2017-2018 NBA season did home teams win when they had 2 nighs
### of rest and the road team was on a B2B
con.execute('''
SELECT home_team, road_team, home_team_final_score, road_team_final_score, road_team_rest, home_team_rest
    FROM nba_schedule
    WHERE road_team_rest = 'B2B'
    AND home_team_rest = '2'
    AND (road_team_final_score < home_team_final_score)
LIMIT 5;
''').fetchall()

[('Memphis', 'Golden State', Decimal('111'), Decimal('101'), 'B2B', '2'),
 ('Oklahoma City', 'Indiana', Decimal('114'), Decimal('96'), 'B2B', '2'),
 ('Denver', 'Brooklyn', Decimal('112'), Decimal('104'), 'B2B', '2'),
 ('Indiana', 'New York', Decimal('115'), Decimal('97'), 'B2B', '2'),
 ('Philadelphia', 'San Antonio', Decimal('112'), Decimal('106'), 'B2B', '2')]

In [42]:
###Query Total
print(len(con.execute('''
SELECT home_team, road_team, home_team_final_score, road_team_final_score, road_team_rest, home_team_rest
    FROM nba_schedule
    WHERE road_team_rest = 'B2B'
    AND home_team_rest = '2'
    AND (road_team_final_score < home_team_final_score)
''').fetchall()))

7


In [43]:
### How many games in the 2017-2018 NBA season did road teams win when they were on a
### 3IN4 and the home team had 1 nights of rest
con.execute('''
SELECT home_team, road_team, home_team_final_score, road_team_final_score, road_team_rest, home_team_rest
    FROM nba_schedule
    WHERE road_team_rest = '3IN4'
    AND home_team_rest = '1'
    AND (road_team_final_score > home_team_final_score)
LIMIT 5;
''').fetchall()

[('Philadelphia', 'Boston', Decimal('92'), Decimal('102'), '3IN4', '1'),
 ('Memphis', 'Orlando', Decimal('99'), Decimal('101'), '3IN4', '1'),
 ('Portland', 'Memphis', Decimal('97'), Decimal('98'), '3IN4', '1'),
 ('Memphis', 'Portland', Decimal('92'), Decimal('100'), '3IN4', '1'),
 ('Dallas', 'Brooklyn', Decimal('104'), Decimal('109'), '3IN4', '1')]

In [44]:
###Query Total
print(len(con.execute('''
SELECT home_team, road_team, home_team_final_score, road_team_final_score, road_team_rest, home_team_rest
    FROM nba_schedule
    WHERE road_team_rest = '3IN4'
    AND home_team_rest = '1'
    AND (road_team_final_score > home_team_final_score)
''').fetchall()))

26


In [45]:
### How many games in the 2017-2018 NBA season did road teams win when they were on a
### 3IN4 and the home team had 1 nights of rest
con.execute('''
SELECT home_team, road_team, home_team_final_score, road_team_final_score, road_team_rest, home_team_rest
    FROM nba_schedule
    WHERE road_team_rest = '3IN4'
    AND home_team_rest = '2'
    AND (road_team_final_score > home_team_final_score)
LIMIT 5;
''').fetchall()

[('Chicago', 'Indiana', Decimal('87'), Decimal('105'), '3IN4', '2'),
 ('LA Clippers', 'Minnesota', Decimal('107'), Decimal('113'), '3IN4', '2'),
 ('Sacramento', 'San Antonio', Decimal('99'), Decimal('108'), '3IN4', '2'),
 ('Miami', 'Brooklyn', Decimal('87'), Decimal('111'), '3IN4', '2'),
 ('Cleveland', 'Golden State', Decimal('108'), Decimal('118'), '3IN4', '2')]

In [46]:
###Query Total
print(len(con.execute('''
SELECT home_team, road_team, home_team_final_score, road_team_final_score, road_team_rest, home_team_rest
    FROM nba_schedule
    WHERE road_team_rest = '3IN4'
    AND home_team_rest = '2'
    AND (road_team_final_score > home_team_final_score)
''').fetchall()))

9


In [47]:
### How many games in the 2017-2018 NBA season did home teams win when they had 1 night of
### rest and the road team was on a B2B
con.execute('''
SELECT home_team, road_team, home_team_final_score, road_team_final_score, road_team_rest, home_team_rest
    FROM nba_schedule
    WHERE road_team_rest = 'B2B'
    AND home_team_rest = '1'
    AND (road_team_final_score < home_team_final_score)
LIMIT 5;
''').fetchall()

[('Phoenix', 'Utah', Decimal('97'), Decimal('88'), 'B2B', '1'),
 ('Washington', 'LA Lakers', Decimal('111'), Decimal('95'), 'B2B', '1'),
 ('Washington', 'Atlanta', Decimal('113'), Decimal('94'), 'B2B', '1'),
 ('New Orleans', 'LA Clippers', Decimal('111'), Decimal('103'), 'B2B', '1'),
 ('Minnesota', 'San Antonio', Decimal('98'), Decimal('86'), 'B2B', '1')]

In [48]:
###Query Total
print(len(con.execute('''
SELECT home_team, road_team, home_team_final_score, road_team_final_score, road_team_rest, home_team_rest
    FROM nba_schedule
    WHERE road_team_rest = '3IN4'
    AND home_team_rest = '1'
    AND (road_team_final_score < home_team_final_score)
''').fetchall()))

37


In [49]:
### How many games in the 2017-2018 NBA season did home teams win when they had 2 nighs of
### rest and the road team was on a 3IN4
con.execute('''
SELECT home_team, road_team, home_team_final_score, road_team_final_score, road_team_rest, home_team_rest
    FROM nba_schedule
    WHERE road_team_rest = '3IN4'
    AND home_team_rest = '2'
    AND (road_team_final_score < home_team_final_score)
LIMIT 5;
''').fetchall()

[('New York', 'Brooklyn', Decimal('107'), Decimal('86'), '3IN4', '2'),
 ('LA Lakers', 'Detroit', Decimal('113'), Decimal('93'), '3IN4', '2'),
 ('Detroit', 'Milwaukee', Decimal('105'), Decimal('96'), '3IN4', '2'),
 ('New York', 'LA Clippers', Decimal('107'), Decimal('85'), '3IN4', '2'),
 ('Cleveland', 'Atlanta', Decimal('123'), Decimal('114'), '3IN4', '2')]

In [50]:
###Query Total
print(len(con.execute('''
SELECT home_team, road_team, home_team_final_score, road_team_final_score, road_team_rest, home_team_rest
    FROM nba_schedule
    WHERE road_team_rest = '3IN4'
    AND home_team_rest = '2'
    AND (road_team_final_score < home_team_final_score)
''').fetchall()))

11


In [51]:
### How many games in the 2017-2018 NBA season did road teams win when they were on a 
### 3IN4-B2B and the road team had at least one night of rest
con.execute('''
SELECT home_team, road_team, home_team_final_score, road_team_final_score, road_team_rest, home_team_rest
    FROM nba_schedule
    WHERE road_team_rest = '3IN4-B2B'
    AND home_team_rest = '1'
    AND (road_team_final_score > home_team_final_score)
LIMIT 5;
''').fetchall()

[('New York', 'Detroit', Decimal('107'), Decimal('111'), '3IN4-B2B', '1'),
 ('Chicago', 'Oklahoma City', Decimal('69'), Decimal('101'), '3IN4-B2B', '1'),
 ('Memphis', 'Charlotte', Decimal('99'), Decimal('104'), '3IN4-B2B', '1'),
 ('New Orleans', 'Orlando', Decimal('99'), Decimal('115'), '3IN4-B2B', '1'),
 ('LA Clippers', 'Golden State', Decimal('113'), Decimal('141'), '3IN4-B2B', '1')]

In [52]:
###Query Total
print(len(con.execute('''
SELECT home_team, road_team, home_team_final_score, road_team_final_score, road_team_rest, home_team_rest
    FROM nba_schedule
    WHERE road_team_rest = '3IN4-B2B'
    AND home_team_rest = '1'
    AND (road_team_final_score > home_team_final_score)
''').fetchall()))

36


In [53]:
### How many games in the 2017-2018 NBA season did home teams win when they had
### at leat 1 night of rest and the road team was on a 3IN4 and B2B
con.execute('''
SELECT home_team, road_team, home_team_final_score, road_team_final_score, road_team_rest, home_team_rest
    FROM nba_schedule
    WHERE road_team_rest = '3IN4-B2B'
    AND home_team_rest = '1'
    AND (road_team_final_score < home_team_final_score)
LIMIT 5;
''').fetchall()

[('Toronto', 'Philadelphia', Decimal('128'), Decimal('94'), '3IN4-B2B', '1'),
 ('LA Clippers', 'Phoenix', Decimal('130'), Decimal('88'), '3IN4-B2B', '1'),
 ('Miami', 'Atlanta', Decimal('104'), Decimal('93'), '3IN4-B2B', '1'),
 ('Detroit', 'Minnesota', Decimal('122'), Decimal('101'), '3IN4-B2B', '1'),
 ('Boston', 'San Antonio', Decimal('108'), Decimal('94'), '3IN4-B2B', '1')]

In [54]:
###Query Total
print(len(con.execute('''
SELECT home_team, road_team, home_team_final_score, road_team_final_score, road_team_rest, home_team_rest
    FROM nba_schedule
    WHERE road_team_rest = '3IN4-B2B'
    AND home_team_rest = '1'
    AND (road_team_final_score < home_team_final_score)
''').fetchall()))

58


In [55]:
### How many games in the 2017-2018 NBA season did road teams win when they were on a 
### 3IN4-B2B and the road team had 2 nights of rest
con.execute('''
SELECT home_team, road_team, home_team_final_score, road_team_final_score, road_team_rest, home_team_rest
    FROM nba_schedule
    WHERE road_team_rest = '3IN4-B2B'
    AND home_team_rest = '2'
    AND (road_team_final_score > home_team_final_score)
LIMIT 5;
''').fetchall()

[('Cleveland', 'Indiana', Decimal('107'), Decimal('124'), '3IN4-B2B', '2'),
 ('Orlando', 'Utah', Decimal('85'), Decimal('125'), '3IN4-B2B', '2'),
 ('Washington', 'Portland', Decimal('105'), Decimal('108'), '3IN4-B2B', '2'),
 ('Brooklyn', 'Sacramento', Decimal('99'), Decimal('104'), '3IN4-B2B', '2'),
 ('Brooklyn', 'Boston', Decimal('85'), Decimal('87'), '3IN4-B2B', '2')]

In [56]:
###Query Total
print(len(con.execute('''
SELECT home_team, road_team, home_team_final_score, road_team_final_score, road_team_rest, home_team_rest
    FROM nba_schedule
    WHERE road_team_rest = '3IN4-B2B'
    AND home_team_rest = '2'
    AND (road_team_final_score > home_team_final_score)
''').fetchall()))

11


In [57]:
### How many games in the 2017-2018 NBA season did home teams win when they had
### at 2 nights of rest and the road team was on a 3IN4 and B2B
con.execute('''
SELECT home_team, road_team, home_team_final_score, road_team_final_score, road_team_rest, home_team_rest
    FROM nba_schedule
    WHERE road_team_rest = '3IN4-B2B'
    AND home_team_rest = '2'
    AND (road_team_final_score < home_team_final_score)
LIMIT 5;
''').fetchall()

[('Miami', 'Indiana', Decimal('112'), Decimal('108'), '3IN4-B2B', '2'),
 ('Houston', 'Dallas', Decimal('107'), Decimal('91'), '3IN4-B2B', '2'),
 ('Denver', 'Sacramento', Decimal('96'), Decimal('79'), '3IN4-B2B', '2'),
 ('Utah', 'LA Lakers', Decimal('96'), Decimal('81'), '3IN4-B2B', '2'),
 ('Minnesota', 'Dallas', Decimal('112'), Decimal('99'), '3IN4-B2B', '2')]

In [58]:
###Query Total
print(len(con.execute('''
SELECT home_team, road_team, home_team_final_score, road_team_final_score, road_team_rest, home_team_rest
    FROM nba_schedule
    WHERE road_team_rest = '3IN4-B2B'
    AND home_team_rest = '2'
    AND (road_team_final_score < home_team_final_score)
''').fetchall()))

26


# <center>Woof, I definitely have query fatigue from that, let's summarize it in a table</center> #

<table>
    <tr> <th> Win Total </th>
         <th> Road Team Rest </th>
         <th> Home Team Rest </th>
         <th> Win Total </th>
    </tr>
    <tr> <th> 15 </th>
         <th> B2B </th>
         <th> 1 </th>
         <th> 31 </th>
    </tr>
    <tr> <th> 5 </th>
         <th> B2B </th>
         <th> 2 </th>
         <th> 7 </th>
    </tr>
    <tr> <th> 26 </th>
         <th> 3IN4 </th>
         <th> 1 </th>
         <th> 37 </th>
    </tr>
    <tr> <th> 9 </th>
         <th> 3IN4 </th>
         <th> 2 </th>
         <th> 11 </th>
    </tr>
    <tr> <th> 36 </th>
         <th> 3IN4-B2B </th>
         <th> 1 </th>
         <th> 58 </th>
    </tr>
    <tr> <th> 11 </th>
         <th> 3IN4-B2B </th>
         <th> 2 </th>
         <th> 26 </th>
    </tr>
    <tr> <th> Sum Total of Road Wins </th>
         <th> ... </th>
         <th> ... </th>
         <th> Sum Total of Home Wins </th>
    </tr>
    <tr> <th> 102 </th>
         <th> ... </th>
         <th> ... </th>
         <th> 170 </th>
    </tr>
</table>

In summary when the home team is in a rest advantagous situation they win 170/272 or 62.5% of the time while when the road team is in a rest disadventageous situation they win 102/272 37.5% of the time

However these situations only make up for 272/1312 or 20% of all NBA games in a given season.

# Ok that was somewhat insightful but let's load up a table that has the total wins from all 30 NBA teams in the 2017-2018 season. I'll use this table along with my NBA cities table that has geo-data to run some more queries to see if travel distance and overall record (i.e. how good a team is, affects performance in NBA games #

Ok first let's read in our CSV file and then load it into a table in our elephant SQL databse

In [337]:
team_wins_and_losses = pd.read_csv("team_wins_and_losses.csv")

In [338]:
team_wins_and_losses.head()

team  wins  losses
0    Atlanta    24      58
1     Boston    55      27
2   Brooklyn    28      54
3  Charlotte    36      46
4    Chicago    27      55

In [339]:
team_wins_and_losses_dict = team_wins_and_losses.to_dict('records')

In [340]:
team_wins_and_losses_dict[0]

{'team': 'Atlanta', 'wins': 24, 'losses': 58}

In [341]:
rs = con.execute('''
DROP TABLE IF EXISTS team_wins_and_losses;
CREATE TABLE team_wins_and_losses
	(
	  team character varying(50),
      wins numeric,
      losses numeric
	);
''')

In [342]:
for row in team_wins_and_losses_dict:
    cmd = sqlalchemy.sql.text('''INSERT INTO team_wins_and_losses(team, wins, losses)\
    VALUES (:team, :wins, :losses) ''' )
    con.execute(cmd, row)

In [345]:
con.execute("SELECT * from team_wins_and_losses LIMIT 5").fetchall()

[('Atlanta', Decimal('24'), Decimal('58')),
 ('Boston', Decimal('55'), Decimal('27')),
 ('Brooklyn', Decimal('28'), Decimal('54')),
 ('Charlotte', Decimal('36'), Decimal('46')),
 ('Chicago', Decimal('27'), Decimal('55'))]

In [346]:
con.execute("SELECT * from nba_cities LIMIT 5").fetchall()

[('Atlanta', Decimal('33.749'), Decimal('94.38799999999999'), '0101000020E610000045B6F3FDD49857401D5A643BDFDF4040'),
 ('Boston', Decimal('42.3601'), Decimal('71.0589'), '0101000020E6100000006F8104C5C3514012A5BDC1172E4540'),
 ('Brooklyn', Decimal('40.6782'), Decimal('73.9442'), '0101000020E61000008638D6C56D7C5240D712F241CF564440'),
 ('Charlotte', Decimal('35.2271'), Decimal('80.8431'), '0101000020E61000007DD0B359F5355440F775E09C119D4140'),
 ('Chicago', Decimal('41.8781'), Decimal('87.6298'), '0101000020E610000055C1A8A44EE855400E4FAF9465F04440')]

# OK our table looks good. Let's do some more queries and table joins #

In [347]:
### How many games in the 2017-2018 NBA season did home teams win when they had
### at leat 1 night of rest and the road team was on a 3IN4 and B2B
con.execute('''
SELECT st_distance(
     a.geog::geography,
     b.geog::geography
  ) *  0.000621371192
FROM nba_cities a, nba_cities b
WHERE a.city = 'Detroit' AND b.city = 'Cleveland'
''').fetchall()

[(90.2812332677382,)]

Ok, this query looks good, when I google the "as the crow flies distance" I also get 90miles. The conversion I did in the above query was to convert my output from meters to a more easily managed miles unit. Ok now onto some more insightful queries.

In [349]:
### How many games in the 2017-2018 NBA season did road teams win when they were on a B2B and traveled more than
### 750 miles from their home state???
con.execute('''
SELECT st_distance(
     a.geog::geography,
     b.geog::geography
  ) *  0.000621371192 AS dist, a.city, b.city
FROM nba_cities a, nba_cities b
ORDER BY dist DESC
LIMIT 5;
''').fetchall()

[(2708.34517398914, 'Miami', 'Portland'),
 (2708.34517398914, 'Portland', 'Miami'),
 (2691.16259869268, 'Golden State', 'Boston'),
 (2691.16259869268, 'Boston', 'Golden State'),
 (2631.42251035223, 'Boston', 'Sacremento')]

# OK let's run some queries now to see how often a road team wins on the road when they've had little rest and the home team was good (I.E. won more than 42 games/was above .500)

In [350]:
con.execute('''
SELECT home_team_final_score, road_team_final_score, home_team, road_team
FROM nba_schedule
JOIN team_wins_and_losses ON nba_schedule.home_team = team_wins_and_losses.team 
WHERE nba_schedule.road_team_rest = 'B2B'
    AND nba_schedule.home_team_rest = '1'
    AND (nba_schedule.road_team_final_score > nba_schedule.home_team_final_score)
    AND team_wins_and_losses.wins > 42
    LIMIT 3;
''').fetchall()

[(Decimal('117'), Decimal('126'), 'Cleveland', 'Denver'),
 (Decimal('107'), Decimal('115'), 'Golden State', 'Detroit'),
 (Decimal('101'), Decimal('109'), 'Miami', 'Houston')]

In [352]:
print(len(con.execute('''
SELECT home_team_final_score, road_team_final_score, home_team, road_team
FROM nba_schedule
JOIN team_wins_and_losses ON nba_schedule.home_team = team_wins_and_losses.team 
WHERE nba_schedule.road_team_rest = 'B2B'
    AND nba_schedule.home_team_rest = '1'
    AND (nba_schedule.road_team_final_score > nba_schedule.home_team_final_score)
    AND team_wins_and_losses.wins > 42
''').fetchall()))

4


In [353]:
con.execute('''
SELECT home_team_final_score, road_team_final_score, home_team, road_team
FROM nba_schedule
JOIN team_wins_and_losses ON nba_schedule.home_team = team_wins_and_losses.team 
WHERE nba_schedule.road_team_rest = 'B2B'
    AND nba_schedule.home_team_rest = '2'
    AND (nba_schedule.road_team_final_score > nba_schedule.home_team_final_score)
    AND team_wins_and_losses.wins > 42
    LIMIT 3;
''').fetchall()

[(Decimal('116'), Decimal('122'), 'Washington', 'Phoenix'),
 (Decimal('92'), Decimal('106'), 'Portland', 'Washington'),
 (Decimal('124'), Decimal('125'), 'Boston', 'Washington')]

In [354]:
print(len(con.execute('''
SELECT home_team_final_score, road_team_final_score, home_team, road_team
FROM nba_schedule
JOIN team_wins_and_losses ON nba_schedule.home_team = team_wins_and_losses.team 
WHERE nba_schedule.road_team_rest = 'B2B'
    AND nba_schedule.home_team_rest = '2'
    AND (nba_schedule.road_team_final_score > nba_schedule.home_team_final_score)
    AND team_wins_and_losses.wins > 42
''').fetchall()))

3


In [356]:
con.execute('''
SELECT home_team_final_score, road_team_final_score, home_team, road_team
FROM nba_schedule
JOIN team_wins_and_losses ON nba_schedule.home_team = team_wins_and_losses.team 
WHERE nba_schedule.road_team_rest = '3IN4'
    AND nba_schedule.home_team_rest = '1'
    AND (nba_schedule.road_team_final_score > nba_schedule.home_team_final_score)
    AND team_wins_and_losses.wins > 42
    LIMIT 3;
''').fetchall()

[(Decimal('103'), Decimal('111'), 'Boston', 'Washington'),
 (Decimal('97'), Decimal('110'), 'Denver', 'Atlanta'),
 (Decimal('91'), Decimal('110'), 'Golden State', 'Utah')]

In [357]:
print(len(con.execute('''
SELECT home_team_final_score, road_team_final_score, home_team, road_team
FROM nba_schedule
JOIN team_wins_and_losses ON nba_schedule.home_team = team_wins_and_losses.team 
WHERE nba_schedule.road_team_rest = '3IN4'
    AND nba_schedule.home_team_rest = '1'
    AND (nba_schedule.road_team_final_score > nba_schedule.home_team_final_score)
    AND team_wins_and_losses.wins > 42
''').fetchall()))

11


In [358]:
con.execute('''
SELECT home_team_final_score, road_team_final_score, home_team, road_team
FROM nba_schedule
JOIN team_wins_and_losses ON nba_schedule.home_team = team_wins_and_losses.team 
WHERE nba_schedule.road_team_rest = '3IN4'
    AND nba_schedule.home_team_rest = '2'
    AND (nba_schedule.road_team_final_score > nba_schedule.home_team_final_score)
    AND team_wins_and_losses.wins > 42
    LIMIT 3;
''').fetchall()

[(Decimal('87'), Decimal('111'), 'Miami', 'Brooklyn'),
 (Decimal('108'), Decimal('118'), 'Cleveland', 'Golden State'),
 (Decimal('88'), Decimal('89'), 'Miami', 'Sacramento')]

In [359]:
print(len(con.execute('''
SELECT home_team_final_score, road_team_final_score, home_team, road_team
FROM nba_schedule
JOIN team_wins_and_losses ON nba_schedule.home_team = team_wins_and_losses.team 
WHERE nba_schedule.road_team_rest = '3IN4'
    AND nba_schedule.home_team_rest = '2'
    AND (nba_schedule.road_team_final_score > nba_schedule.home_team_final_score)
    AND team_wins_and_losses.wins > 42
''').fetchall()))

5


In [360]:
con.execute('''
SELECT home_team_final_score, road_team_final_score, home_team, road_team
FROM nba_schedule
JOIN team_wins_and_losses ON nba_schedule.home_team = team_wins_and_losses.team 
WHERE nba_schedule.road_team_rest = '3IN4-B2B'
    AND nba_schedule.home_team_rest = '1'
    AND (nba_schedule.road_team_final_score > nba_schedule.home_team_final_score)
    AND team_wins_and_losses.wins > 42
    LIMIT 3;
''').fetchall()

[(Decimal('120'), Decimal('126'), 'Golden State', 'New Orleans'),
 (Decimal('95'), Decimal('118'), 'Indiana', 'Houston'),
 (Decimal('98'), Decimal('104'), 'Indiana', 'Detroit')]

In [361]:
print(len(con.execute('''
SELECT home_team_final_score, road_team_final_score, home_team, road_team
FROM nba_schedule
JOIN team_wins_and_losses ON nba_schedule.home_team = team_wins_and_losses.team 
WHERE nba_schedule.road_team_rest = '3IN4-B2B'
    AND nba_schedule.home_team_rest = '1'
    AND (nba_schedule.road_team_final_score > nba_schedule.home_team_final_score)
    AND team_wins_and_losses.wins > 42
''').fetchall()))

12


In [362]:
con.execute('''
SELECT home_team_final_score, road_team_final_score, home_team, road_team
FROM nba_schedule
JOIN team_wins_and_losses ON nba_schedule.home_team = team_wins_and_losses.team 
WHERE nba_schedule.road_team_rest = '3IN4-B2B'
    AND nba_schedule.home_team_rest = '2'
    AND (nba_schedule.road_team_final_score > nba_schedule.home_team_final_score)
    AND team_wins_and_losses.wins > 42
    LIMIT 3;
''').fetchall()

[(Decimal('107'), Decimal('124'), 'Cleveland', 'Indiana'),
 (Decimal('94'), Decimal('109'), 'Utah', 'Indiana'),
 (Decimal('105'), Decimal('108'), 'Washington', 'Portland')]

In [363]:
print(len(con.execute('''
SELECT home_team_final_score, road_team_final_score, home_team, road_team
FROM nba_schedule
JOIN team_wins_and_losses ON nba_schedule.home_team = team_wins_and_losses.team 
WHERE nba_schedule.road_team_rest = '3IN4-B2B'
    AND nba_schedule.home_team_rest = '2'
    AND (nba_schedule.road_team_final_score > nba_schedule.home_team_final_score)
    AND team_wins_and_losses.wins > 42
''').fetchall()))

3


# Now let's do the reverse of this, how often does the home team win when they are good, and rested, AND the road team is not well rested #

In [364]:
con.execute('''
SELECT home_team_final_score, road_team_final_score, home_team, road_team
FROM nba_schedule
JOIN team_wins_and_losses ON nba_schedule.home_team = team_wins_and_losses.team 
WHERE nba_schedule.road_team_rest = 'B2B'
    AND nba_schedule.home_team_rest = '1'
    AND (nba_schedule.road_team_final_score < nba_schedule.home_team_final_score)
    AND team_wins_and_losses.wins > 42
    LIMIT 3;
''').fetchall()

[(Decimal('103'), Decimal('73'), 'Boston', 'New York'),
 (Decimal('121'), Decimal('104'), 'Cleveland', 'Detroit'),
 (Decimal('129'), Decimal('123'), 'Cleveland', 'Brooklyn')]

In [365]:
print(len(con.execute('''
SELECT home_team_final_score, road_team_final_score, home_team, road_team
FROM nba_schedule
JOIN team_wins_and_losses ON nba_schedule.home_team = team_wins_and_losses.team 
WHERE nba_schedule.road_team_rest = 'B2B'
    AND nba_schedule.home_team_rest = '1'
    AND (nba_schedule.road_team_final_score < nba_schedule.home_team_final_score)
    AND team_wins_and_losses.wins > 42
''').fetchall()))

23


In [366]:
con.execute('''
SELECT home_team_final_score, road_team_final_score, home_team, road_team
FROM nba_schedule
JOIN team_wins_and_losses ON nba_schedule.home_team = team_wins_and_losses.team 
WHERE nba_schedule.road_team_rest = 'B2B'
    AND nba_schedule.home_team_rest = '2'
    AND (nba_schedule.road_team_final_score < nba_schedule.home_team_final_score)
    AND team_wins_and_losses.wins > 42
    LIMIT 3;
''').fetchall()

[(Decimal('114'), Decimal('96'), 'Oklahoma City', 'Indiana'),
 (Decimal('112'), Decimal('104'), 'Denver', 'Brooklyn'),
 (Decimal('115'), Decimal('97'), 'Indiana', 'New York')]

In [367]:
print(len(con.execute('''
SELECT home_team_final_score, road_team_final_score, home_team, road_team
FROM nba_schedule
JOIN team_wins_and_losses ON nba_schedule.home_team = team_wins_and_losses.team 
WHERE nba_schedule.road_team_rest = 'B2B'
    AND nba_schedule.home_team_rest = '2'
    AND (nba_schedule.road_team_final_score < nba_schedule.home_team_final_score)
    AND team_wins_and_losses.wins > 42
''').fetchall()))

6


In [368]:
con.execute('''
SELECT home_team_final_score, road_team_final_score, home_team, road_team
FROM nba_schedule
JOIN team_wins_and_losses ON nba_schedule.home_team = team_wins_and_losses.team 
WHERE nba_schedule.road_team_rest = '3IN4'
    AND nba_schedule.home_team_rest = '1'
    AND (nba_schedule.road_team_final_score < nba_schedule.home_team_final_score)
    AND team_wins_and_losses.wins > 42
    LIMIT 3;
''').fetchall()

[(Decimal('112'), Decimal('90'), 'Cleveland', 'Detroit'),
 (Decimal('128'), Decimal('125'), 'Denver', 'Milwaukee'),
 (Decimal('110'), Decimal('100'), 'Golden State', 'Orlando')]

In [369]:
print(len(con.execute('''
SELECT home_team_final_score, road_team_final_score, home_team, road_team
FROM nba_schedule
JOIN team_wins_and_losses ON nba_schedule.home_team = team_wins_and_losses.team 
WHERE nba_schedule.road_team_rest = '3IN4'
    AND nba_schedule.home_team_rest = '1'
    AND (nba_schedule.road_team_final_score < nba_schedule.home_team_final_score)
    AND team_wins_and_losses.wins > 42
''').fetchall()))

28


In [370]:
con.execute('''
SELECT home_team_final_score, road_team_final_score, home_team, road_team
FROM nba_schedule
JOIN team_wins_and_losses ON nba_schedule.home_team = team_wins_and_losses.team 
WHERE nba_schedule.road_team_rest = '3IN4'
    AND nba_schedule.home_team_rest = '2'
    AND (nba_schedule.road_team_final_score < nba_schedule.home_team_final_score)
    AND team_wins_and_losses.wins > 42
    LIMIT 3;
''').fetchall()

[(Decimal('123'), Decimal('114'), 'Cleveland', 'Atlanta'),
 (Decimal('119'), Decimal('116'), 'Milwaukee', 'Cleveland'),
 (Decimal('102'), Decimal('96'), 'Washington', 'Oklahoma City')]

In [371]:
print(len(con.execute('''
SELECT home_team_final_score, road_team_final_score, home_team, road_team
FROM nba_schedule
JOIN team_wins_and_losses ON nba_schedule.home_team = team_wins_and_losses.team 
WHERE nba_schedule.road_team_rest = '3IN4'
    AND nba_schedule.home_team_rest = '2'
    AND (nba_schedule.road_team_final_score < nba_schedule.home_team_final_score)
    AND team_wins_and_losses.wins > 42
''').fetchall()))

5


In [374]:
con.execute('''
SELECT home_team_final_score, road_team_final_score, home_team, road_team
FROM nba_schedule
JOIN team_wins_and_losses ON nba_schedule.home_team = team_wins_and_losses.team 
WHERE nba_schedule.road_team_rest = '3IN4-B2B'
    AND nba_schedule.home_team_rest = '1'
    AND (nba_schedule.road_team_final_score < nba_schedule.home_team_final_score)
    AND team_wins_and_losses.wins > 42
    LIMIT 3;
''').fetchall()

[(Decimal('108'), Decimal('94'), 'Boston', 'San Antonio'),
 (Decimal('113'), Decimal('86'), 'Boston', 'Sacramento'),
 (Decimal('134'), Decimal('106'), 'Boston', 'Charlotte')]

In [375]:
print(len(con.execute('''
SELECT home_team_final_score, road_team_final_score, home_team, road_team
FROM nba_schedule
JOIN team_wins_and_losses ON nba_schedule.home_team = team_wins_and_losses.team 
WHERE nba_schedule.road_team_rest = '3IN4-B2B'
    AND nba_schedule.home_team_rest = '1'
    AND (nba_schedule.road_team_final_score < nba_schedule.home_team_final_score)
    AND team_wins_and_losses.wins > 42
''').fetchall()))

39


In [376]:
con.execute('''
SELECT home_team_final_score, road_team_final_score, home_team, road_team
FROM nba_schedule
JOIN team_wins_and_losses ON nba_schedule.home_team = team_wins_and_losses.team 
WHERE nba_schedule.road_team_rest = '3IN4-B2B'
    AND nba_schedule.home_team_rest = '2'
    AND (nba_schedule.road_team_final_score < nba_schedule.home_team_final_score)
    AND team_wins_and_losses.wins > 42
    LIMIT 3;
''').fetchall()

[(Decimal('108'), Decimal('97'), 'Boston', 'Philadelphia'),
 (Decimal('127'), Decimal('110'), 'Cleveland', 'Portland'),
 (Decimal('96'), Decimal('79'), 'Denver', 'Sacramento')]

In [377]:
print(len(con.execute('''
SELECT home_team_final_score, road_team_final_score, home_team, road_team
FROM nba_schedule
JOIN team_wins_and_losses ON nba_schedule.home_team = team_wins_and_losses.team 
WHERE nba_schedule.road_team_rest = '3IN4-B2B'
    AND nba_schedule.home_team_rest = '2'
    AND (nba_schedule.road_team_final_score < nba_schedule.home_team_final_score)
    AND team_wins_and_losses.wins > 42
''').fetchall()))

20


# Let's summarize this with a table like before, only now in this case the home team was good AND well rested #

<table>
    <tr> <th> Win Total </th>
         <th> Road Team Rest </th>
         <th> Home Team Rest </th>
         <th> Win Total </th>
    </tr>
    <tr> <th> 4 </th>
         <th> B2B </th>
         <th> 1 </th>
         <th> 23 </th>
    </tr>
    <tr> <th> 3 </th>
         <th> B2B </th>
         <th> 2 </th>
         <th> 6 </th>
    </tr>
    <tr> <th> 11 </th>
         <th> 3IN4 </th>
         <th> 1 </th>
         <th> 28 </th>
    </tr>
    <tr> <th> 5 </th>
         <th> 3IN4 </th>
         <th> 2 </th>
         <th> 5 </th>
    </tr>
    <tr> <th> 12 </th>
         <th> 3IN4-B2B </th>
         <th> 1 </th>
         <th> 39 </th>
    </tr>
    <tr> <th> 3 </th>
         <th> 3IN4-B2B </th>
         <th> 2 </th>
         <th> 20 </th>
    </tr>
    <tr> <th> Sum Total of Road Wins </th>
         <th> ... </th>
         <th> ... </th>
         <th> Sum Total of Home Wins </th>
    </tr>
    <tr> <th> 38 </th>
         <th> ... </th>
         <th> ... </th>
         <th> 121 </th>
    </tr>
</table>

# Some final fun and simple Queries would be to compute the distance between NBA cities. I didn't use this for my analysis of game outcomes but it would definitely be in the scope of the next iteration/piece of this project. #

In [17]:
### Query to return the 2 farthest apart NBA cities
con.execute('''
SELECT st_distance(
     a.geog::geography,
     b.geog::geography
  ) *  0.000621371192 AS dist, a.city, b.city
FROM nba_cities a, nba_cities b
ORDER BY DIST DESC
LIMIT 1;
''').fetchall()

[(2708.34517398914, 'Miami', 'Portland')]

In [32]:
### Query to return the 2 closest NBA cities
con.execute('''
SELECT st_distance(
     a.geog::geography,
     b.geog::geography
  ) *  0.000621371192 AS dist, a.city, b.city
FROM nba_cities a, nba_cities b
WHERE ST_Distance(a.geog, b.geog) > 0
ORDER BY dist ASC
LIMIT 1;
''').fetchall()

[(4.42538323382058, 'Brooklyn', 'New York')]

In [16]:
con.execute('''
SELECT *
FROM
(SELECT st_distance(
     a.geog::geography,
     b.geog::geography
  ) AS dist, a.city AS acity, b.city AS bcity
    FROM nba_cities a, nba_cities b
    WHERE st_distance(a.geog, b.geog) > 3000000.0000000000
) t1
INNER JOIN
(
    SELECT home_team_final_score, road_team_final_score, home_team, road_team
    FROM nba_schedule 
    WHERE nba_schedule.road_team_rest = '3IN4-B2B'
    AND nba_schedule.home_team_rest = '1'
    AND (nba_schedule.road_team_final_score < nba_schedule.home_team_final_score)) t2
ON t2.home_team = t1.acity AND t2.road_team = t1.bcity
''').fetchall()

[(4163595.9106841, 'Golden State', 'Miami', Decimal('97'), Decimal('80'), 'Golden State', 'Miami'),
 (3501566.82746384, 'LA Clippers', 'Toronto', Decimal('96'), Decimal('91'), 'LA Clippers', 'Toronto'),
 (3169960.17898927, 'Detroit', 'Portland', Decimal('111'), Decimal('91'), 'Detroit', 'Portland'),
 (3189657.90243239, 'Miami', 'Phoenix', Decimal('125'), Decimal('103'), 'Miami', 'Phoenix')]

In [ ]:
SELECT home_team_final_score, road_team_final_score, home_team, road_team
FROM nba_schedule
JOIN team_wins_and_losses ON nba_schedule.home_team = team_wins_and_losses.team 
WHERE nba_schedule.road_team_rest = '3IN4-B2B'
    AND nba_schedule.home_team_rest = '1'
    AND (nba_schedule.road_team_final_score < nba_schedule.home_team_final_score)
    AND team_wins_and_losses.wins > 42
    LIMIT 3;

In [202]:
### How many games in the 2017-2018 NBA season did road teams win when they were on a B2B and traveled more than
### 750 miles from their home state???
con.execute('''
SELECT ST_DWithin(
     a.geog::geography,
     b.geog::geography,
     1000000
  ), a.city, b.city
FROM nba_cities a, nba_cities b
LIMIT 5;
''').fetchall()

[(True, 'Atlanta', 'Atlanta'),
 (False, 'Atlanta', 'Boston'),
 (False, 'Atlanta', 'Brooklyn'),
 (False, 'Atlanta', 'Charlotte'),
 (False, 'Atlanta', 'Chicago')]

In [194]:
con.execute('''
SELECT home_team_final_score, road_team_final_score, home_team, road_team
FROM nba_schedule
WHERE road_team_rest = '3IN4-B2B'
    AND home_team_rest = '1'
    AND (road_team_final_score > home_team_final_score)
    LIMIT 10;
''').fetchall()

[(Decimal('107'), Decimal('111'), 'New York', 'Detroit'),
 (Decimal('69'), Decimal('101'), 'Chicago', 'Oklahoma City'),
 (Decimal('99'), Decimal('104'), 'Memphis', 'Charlotte'),
 (Decimal('99'), Decimal('115'), 'New Orleans', 'Orlando'),
 (Decimal('113'), Decimal('141'), 'LA Clippers', 'Golden State'),
 (Decimal('95'), Decimal('118'), 'Indiana', 'Houston'),
 (Decimal('116'), Decimal('125'), 'New Orleans', 'Toronto'),
 (Decimal('111'), Decimal('118'), 'Brooklyn', 'Golden State'),
 (Decimal('97'), Decimal('105'), 'Orlando', 'Indiana'),
 (Decimal('88'), Decimal('99'), 'Milwaukee', 'Washington')]

In [301]:
### How many games in the 2017-2018 NBA season did road teams win when they were on a B2B and traveled more than
### 750 miles from their home state???
con.execute('''
SELECT st_distance(
     a.geog::geography,
     b.geog::geography
  ) *  0.000621371192 AS dist, a.city, b.city
FROM nba_cities a, nba_cities b
ORDER BY dist DESC
LIMIT 5;
''').fetchall()

[(2708.34517398914, 'Miami', 'Portland'),
 (2708.34517398914, 'Portland', 'Miami'),
 (2691.16259869268, 'Golden State', 'Boston'),
 (2691.16259869268, 'Boston', 'Golden State'),
 (2631.42251035223, 'Boston', 'Sacremento')]